### DISCUSSION TITLE
celery worker timed out with rabbitmq

### DISCUSSION THREAD

**Discussion by hadoopBrogrammers** [apache/airflow#27350] [General]
airflow version 2.0.2
rabbitmqctl version: 3.8.16
celery worker version : 4.4.7 (cliffs)
airflow installed on Kubernetes ..

1. It is working fine but after some run worker node got stuck with following error ....
  File "/home/airflow/.local/lib/python3.6/site-packages/amqp/channel.py", line 280, in _on_close
    reply_code, reply_text, (class_id, method_id), ChannelError,
amqp.exceptions.PreconditionFailed: (0, 0): (406) PRECONDITION_FAILED - consumer ack timed out on channel 1

2. When I checked Rabitmq log then I see following error....
2021-05-12 10:31:13.882 [info] <0.19109.0> Closing all channels from connection '*.*.*.216:37624 -> *.*.*.214:5672' because it has been closed
2021-05-12 10:31:14.135 [info] <0.19114.0> accepting AMQP connection <0.19114.0> (*.*.*.216:37644 -> *.*.*.214:5672)
2021-05-12 10:31:14.137 [info] <0.19114.0> connection <0.19114.0> (*.*.*.216:37644 -> *.*.*.214:5672): user 'airflow' authenticated and granted access to vhost 'airflow'
2021-05-12 12:15:01.523 [warning] <0.2188.0> Consumer None4 on channel 1 has timed out waiting on consumer acknowledgement. Timeout used: 900000 ms
2021-05-12 12:15:01.525 [error] <0.2188.0> Channel error on connection <0.2173.0> (*.*.*.217:51938 -> *.*.*.214:5672, vhost: 'airflow', user: 'airflow'), channel 1:
operation none caused a channel exception precondition_failed: consumer ack timed out on channel 1
2021-05-12 12:17:33.050 [error] <0.2173.0> closing AMQP connection <0.2173.0> (*.*.*.217:51938 -> *.*.*.214:5672):
missed heartbeats from client, timeout: 60s
2021-05-12 12:17:33.050 [info] <0.20579.0> Closing all channels from connection '*.*.*.217:51938 -> *.*.*.214:5672' because it has been closed

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by rcwoolston** [comment]
It is due to this change https://github.com/rabbitmq/rabbitmq-server/pull/2990. It would be nice if you could set a global default timeout in the airflow.cfg that matched your maximum allowed for your consumers.

**Comment by Danirill** [comment]
Hello, i don't have airflow, but got the same error. There is my django app log:
```
[2021-06-05 15:30:08,313: DEBUG/MainProcess] Channel open
[2021-06-05 15:30:08,314: CRITICAL/MainProcess] Unrecoverable error: PreconditionFailed(406, 'PRECONDITION_FAILED - consumer ack timed out on channel 1', (0, 0), '')
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/celery/worker/worker.py", line 203, in start
    self.blueprint.start(self)
  File "/usr/local/lib/python3.7/site-packages/celery/bootsteps.py", line 116, in start
    step.start(parent)
  File "/usr/local/lib/python3.7/site-packages/celery/bootsteps.py", line 365, in start
    return self.obj.start()
  File "/usr/local/lib/python3.7/site-packages/celery/worker/consumer/consumer.py", line 326, in start
    blueprint.start(self)
  File "/usr/local/lib/python3.7/site-packages/celery/bootsteps.py", line 116, in start
    step.start(parent)
  File "/usr/local/lib/python3.7/site-packages/celery/worker/consumer/consumer.py", line 618, in start
    c.loop(*c.loop_args())
  File "/usr/local/lib/python3.7/site-packages/celery/worker/loops.py", line 81, in asynloop
    next(loop)
  File "/usr/local/lib/python3.7/site-packages/kombu/asynchronous/hub.py", line 361, in create_loop
    cb(*cbargs)
  File "/usr/local/lib/python3.7/site-packages/kombu/transport/base.py", line 235, in on_readable
    reader(loop)
  File "/usr/local/lib/python3.7/site-packages/kombu/transport/base.py", line 217, in _read
    drain_events(timeout=0)
  File "/usr/local/lib/python3.7/site-packages/amqp/connection.py", line 523, in drain_events
    while not self.blocking_read(timeout):
  File "/usr/local/lib/python3.7/site-packages/amqp/connection.py", line 529, in blocking_read
    return self.on_inbound_frame(frame)
  File "/usr/local/lib/python3.7/site-packages/amqp/method_framing.py", line 53, in on_frame
    callback(channel, method_sig, buf, None)
  File "/usr/local/lib/python3.7/site-packages/amqp/connection.py", line 536, in on_inbound_method
    method_sig, payload, content,
  File "/usr/local/lib/python3.7/site-packages/amqp/abstract_channel.py", line 143, in dispatch_method
    listener(*args)
  File "/usr/local/lib/python3.7/site-packages/amqp/channel.py", line 278, in _on_close
    reply_code, reply_text, (class_id, method_id), ChannelError,
amqp.exceptions.PreconditionFailed: (0, 0): (406) PRECONDITION_FAILED - consumer ack timed out on channel 1
[2021-06-05 15:30:08,352: DEBUG/MainProcess] Closed channel #3
[2021-06-05 15:30:08,408: DEBUG/MainProcess] Closed channel #2
[2021-06-05 15:30:09,474: DEBUG/MainProcess] Closed channel #1
[2021-06-05 15:30:09,481: DEBUG/MainProcess] Closed channel #1
```
Error repeats periodically. I tried to solve this problem by aading this code to settings.py:
```
BROKER_POOL_LIMIT = 10
CELERYD_FORCE_EXECV = True
CELERY_TASK_RESULT_EXPIRES = 3600
CELERYD_POOL_RESTARTS = True
CELERY_ACKS_LATE = True
CELERYD_PREFETCH_MULTIPLIER = 1
```
But it gave no results.
Thanks

**Comment by enanablancaynumeros** [comment]
I have the same issue after upgrading the rabbitmq cluster to 3.8.16 and using celery 5.1.1.

As reported, it seems to be caused by the broker change https://github.com/rabbitmq/rabbitmq-server/pull/2990 but I can't find any option to control from celery the broker consumer_timeout.

**Comment by potiuk** [comment]
> I have the same issue after upgrading the rabbitmq cluster to 3.8.16 and using celery 5.1.1.

Note that Celery 5 is not YET officially supported by Airflow. There is an open issue for that #11301 - and if you would like to make a PR to add support for it, you are most welcome!

From setup.py:

```
    'celery~=4.4.2',
```

**Comment by enanablancaynumeros** [comment]
> Note that Celery 5 is not YET officially supported by Airflow. There is an open issue for that #11301 - and if you would like to make a PR to add support for it, you are most welcome!
> 
> From setup.py:
> 
> ```
>     'celery~=4.4.2',
> ```

My bad, I just realized that this is the airflow project and not the celery one :sweat_smile:

**Comment by potiuk** [comment]
> My bad, I just realized that this is the airflow project and not the celery one 😅

Been there, done that  :D

**Comment by huozhanfeng** [comment]
We get the same error on version airflow[2.1.0]

**Comment by ubearcat** [comment]
We are seeing this issue with Airflow 1.10.14 and celery 4.3

**Comment by huozhanfeng** [comment]
> We get the same error on version airflow[2.1.0]

we can turn params 'consumer_timeout' to solve the problem when the version of the RabbitMQ  >= 3.8.15. And here is a discussion for it, https://stackoverflow.com/questions/67907336/celery-task-with-a-long-eta-and-rabbitmq

**Comment by huozhanfeng** [comment]
> I have the same issue after upgrading the rabbitmq cluster to 3.8.16 and using celery 5.1.1.
> 
> As reported, it seems to be caused by the broker change [rabbitmq/rabbitmq-server#2990](https://github.com/rabbitmq/rabbitmq-server/pull/2990) but I can't find any option to control from celery the broker consumer_timeout.

It works, thanks a lot.

**Comment by noelmcloughlin** [comment]
Link to RMQ consumer_timeout: https://www.rabbitmq.com/consumers.html#acknowledgement-timeout

**Comment by eladkal** [comment]
This issue is reported against older version of Airflow.
Please check with latest Airflow version.

**Comment by theophanevie** [comment]
Having this issue on airflow `2.4.1` aswell

celery version 5.2.7

celery configuration : 
``` python
celery_app.conf.update(
    worker_prefetch_multiplier=1,
    task_acks_late=True,
    task_reject_on_worker_lost=True,
    worker_max_tasks_per_child=100,
    worker_cancel_long_running_tasks_on_connection_loss=True,
)
```

worker log :
``` shell
[2022-10-20T06:16:28.195699315Z] ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:2396)
--
  |   | [2022-10-20T06:16:28.195695965Z] return self._sslobj.write(data)
  |   | [2022-10-20T06:16:28.195692435Z] File "/usr/local/lib/python3.10/ssl.py", line 1149, in write
  |   | [2022-10-20T06:16:28.195689285Z] n = write(s)
  |   | [2022-10-20T06:16:28.195685375Z] File "/usr/local/lib/python3.10/site-packages/amqp/transport.py", line 595, in _write
  |   | [2022-10-20T06:16:28.195681955Z] self._write(s)
  |   | [2022-10-20T06:16:28.195678275Z] File "/usr/local/lib/python3.10/site-packages/amqp/transport.py", line 347, in write
  |   | [2022-10-20T06:16:28.195674215Z] write(buffer_store.view[:offset])
  |   | [2022-10-20T06:16:28.195670455Z] File "/usr/local/lib/python3.10/site-packages/amqp/method_framing.py", line 186, in write_frame
  |   | [2022-10-20T06:16:28.195667145Z] conn.frame_writer(1, self.channel_id, sig, args, content)
  |   | [2022-10-20T06:16:28.195663465Z] File "/usr/local/lib/python3.10/site-packages/amqp/abstract_channel.py", line 70, in send_method
  |   | [2022-10-20T06:16:28.195658615Z] return self.send_method(
  |   | [2022-10-20T06:16:28.195655015Z] File "/usr/local/lib/python3.10/site-packages/amqp/channel.py", line 1407, in basic_ack
  |   | [2022-10-20T06:16:28.195651455Z] self.channel.basic_ack(self.delivery_tag, multiple=multiple)
  |   | [2022-10-20T06:16:28.195643425Z] File "/usr/local/lib/python3.10/site-packages/kombu/message.py", line 123, in ack
  |   | [2022-10-20T06:16:28.195640275Z] self.ack(multiple=multiple)
  |   | [2022-10-20T06:16:28.195636695Z] File "/usr/local/lib/python3.10/site-packages/kombu/message.py", line 128, in ack_log_error
  |   | [2022-10-20T06:16:28.195633284Z] Traceback (most recent call last):
  |   | [2022-10-20T06:16:28.195630094Z]
  |   | [2022-10-20T06:16:28.195626734Z] During handling of the above exception, another exception occurred:
  |   | [2022-10-20T06:16:28.195623684Z]
  |   | [2022-10-20T06:16:28.195619844Z] amqp.exceptions.PreconditionFailed: (0, 0): (406) PRECONDITION_FAILED - delivery acknowledgement on channel 1 timed out. Timeout value used: 1800000 ms. This timeout value can be configured, see consumers doc guide to learn more
  |   | [2022-10-20T06:16:28.195616414Z] raise error_for_code(
  |   | [2022-10-20T06:16:28.195612414Z] File "/usr/local/lib/python3.10/site-packages/amqp/channel.py", line 293, in _on_close
  |   | [2022-10-20T06:16:28.195609064Z] listener(*args)
  |   | [2022-10-20T06:16:28.195604304Z] File "/usr/local/lib/python3.10/site-packages/amqp/abstract_channel.py", line 156, in dispatch_method
  |   | [2022-10-20T06:16:28.195600954Z] return self.channels[channel_id].dispatch_method(
  |   | [2022-10-20T06:16:28.195597284Z] File "/usr/local/lib/python3.10/site-packages/amqp/connection.py", line 537, in on_inbound_method
  |   | [2022-10-20T06:16:28.195593984Z] callback(channel, method_sig, buf, None)
  |   | [2022-10-20T06:16:28.195590474Z] File "/usr/local/lib/python3.10/site-packages/amqp/method_framing.py", line 53, in on_frame
  |   | [2022-10-20T06:16:28.195587124Z] return self.on_inbound_frame(frame)
  |   | [2022-10-20T06:16:28.195583494Z] File "/usr/local/lib/python3.10/site-packages/amqp/connection.py", line 531, in blocking_read
  |   | [2022-10-20T06:16:28.195580234Z] while not self.blocking_read(timeout):
  |   | [2022-10-20T06:16:28.195576514Z] File "/usr/local/lib/python3.10/site-packages/amqp/connection.py", line 525, in drain_events
  |   | [2022-10-20T06:16:28.195573244Z] drain_events(timeout=0)
  |   | [2022-10-20T06:16:28.195569814Z] File "/usr/local/lib/python3.10/site-packages/kombu/transport/base.py", line 217, in _read
  |   | [2022-10-20T06:16:28.195566844Z] reader(loop)
  |   | [2022-10-20T06:16:28.195563264Z] File "/usr/local/lib/python3.10/site-packages/kombu/transport/base.py", line 235, in on_readable
  |   | [2022-10-20T06:16:28.195560104Z] cb(*cbargs)
  |   | [2022-10-20T06:16:28.195556484Z] File "/usr/local/lib/python3.10/site-packages/kombu/asynchronous/hub.py", line 362, in create_loop
  |   | [2022-10-20T06:16:28.195553264Z] next(loop)
  |   | [2022-10-20T06:16:28.195549584Z] File "/usr/local/lib/python3.10/site-packages/celery/worker/loops.py", line 97, in asynloop
  |   | [2022-10-20T06:16:28.195546274Z] c.loop(*c.loop_args())
  |   | [2022-10-20T06:16:28.195542404Z] File "/usr/local/lib/python3.10/site-packages/celery/worker/consumer/consumer.py", line 628, in start
  |   | [2022-10-20T06:16:28.195530283Z] step.start(parent)
  |   | [2022-10-20T06:16:28.195526623Z] File "/usr/local/lib/python3.10/site-packages/celery/bootsteps.py", line 116, in start
  |   | [2022-10-20T06:16:28.195523193Z] blueprint.start(self)
  |   | [2022-10-20T06:16:28.195519783Z] File "/usr/local/lib/python3.10/site-packages/celery/worker/consumer/consumer.py", line 332, in start
  |   | [2022-10-20T06:16:28.195516343Z] return self.obj.start()
  |   | [2022-10-20T06:16:28.195512243Z] File "/usr/local/lib/python3.10/site-packages/celery/bootsteps.py", line 365, in start
  |   | [2022-10-20T06:16:28.195488873Z] step.start(parent)
  |   | [2022-10-20T06:16:28.195485193Z] File "/usr/local/lib/python3.10/site-packages/celery/bootsteps.py", line 116, in start
  |   | [2022-10-20T06:16:28.195481333Z] self.blueprint.start(self)
  |   | [2022-10-20T06:16:28.195477453Z] File "/usr/local/lib/python3.10/site-packages/celery/worker/worker.py", line 203, in start
  |   | [2022-10-20T06:16:28.195473203Z] Traceback (most recent call last):
  |   | [2022-10-20T06:16:28.195455983Z] [2022-10-20 06:16:28,193: CRITICAL/MainProcess] Couldn't ack 6, reason:SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:2396)')
```

**Comment by potiuk** [comment]
@theophanevie Seems this might be simply problem with configuration of your rabbitmq not matching the load/characteristics of your traffic (which might be either too low and the problem is with idle connections or too high when cellery workers/rabbitmq cannot cope with the load.

Can you please try the solutions decribed here:

* https://groups.google.com/g/celery-users/c/5dPhSfF5APQ

and possibly here:

* https://stackoverflow.com/questions/52072305/celery-worker-stops-after-being-idle-for-a-few-hours ? 

It seems that might at least reduce the frequency of similar problems

Also it woudl be great if you could describe what kind of setup, sizing, workload characteristics you have and tell us more about  how frequently, in what circumstances it happens. Without having that knowledge  - and just seeing "timeout" message, it next to impossible even to guide you in the direction of how you can solve your problem - which is again, most likely something that needs reconfiguring your rabbitmq. 

Typically this is done by those who manage their deployments via looking at the load and reconfiguring their components as needed - there are no specific quidelines from Airflow on how to do it as every installation has its own set of characteristics and needs indivdual tuning. So you absolutely need to make such an exercise on your load and performance and try to correlate that with the errors you see - before you attempt to fix it.

If you provide the description and circumstances and some kind of result of investigation of what triggers the problem we here might be able to help to tune the setting - but without that, I am afraid it will be difficiult to help to solve your problem.

And BTW. I am converting that into a discussion - because this is most likely question of rabbitmq configuration not Airflow itsefl.

**Comment by anurag1107** [comment]
While configuring Airflow to use RabbitMQ, set the value as follows in airflow.cfg:
-
[celery_broker_transport_options]  consumer_timeout = 31622400000
